source/common/router/router.cc

```diff
@@ -997,13 +1003,34 @@ void Filter::resetOtherUpstreams(UpstreamRequest& upstream_request) {
   final_upstream_request->moveIntoList(std::move(final_upstream_request), upstream_requests_);
 }
 
+void Filter::maybeSetGrpcStatusToOutlierDetection(
+    UpstreamRequest& upstream_request, absl::optional<Grpc::Status::GrpcStatus> grpc_status) {
+  if (grpc_status) {
+    grpc_status_code_received_ = true;
+    if (Http::CodeUtility::is5xx(Grpc::Utility::grpcToHttpStatus(grpc_status.value()))) {
+      // For gRPC request, use the grpc status code for outlier detector when the mapped
+      // http code is 5xx.
+      upstream_request.upstream_host_->outlierDetector().putHttpResponseCode(
+          Grpc::Utility::grpcToHttpStatus(grpc_status.value()));
+    } else {
+      upstream_request.upstream_host_->outlierDetector().putHttpResponseCode(http_status_code_);
+    }
+  }
+}
+
 void Filter::onUpstreamHeaders(uint64_t response_code, Http::HeaderMapPtr&& headers,
                                UpstreamRequest& upstream_request, bool end_stream) {
   ENVOY_STREAM_LOG(debug, "upstream headers complete: end_stream={}", *callbacks_, end_stream);
 
   modify_headers_(*headers);
 
-  upstream_request.upstream_host_->outlierDetector().putHttpResponseCode(response_code);
+  if (grpc_request_) {
+    http_status_code_ = response_code;
```

Comment:
Where is "http_status_code" set before this PR? If the field was never set, you can use http_status_code_ (!=0) to indicate if status_code is received for grpc or http, i.e. to save grpc_status_code_received_

Reply:
Done.